gssapi/WrapToken.go

```diff
@@ -0,0 +1,230 @@
+          2        Flags     Attributes field, as described in section
+                             4.2.2.
+          3        Filler    Contains the hex value FF.
+          4..5     EC        Contains the "extra count" field, in big-
+                             endian order as described in section 4.2.3.
+          6..7     RRC       Contains the "right rotation count" in big-
+                             endian order, as described in section
+                             4.2.5.
+          8..15    SND_SEQ   Sequence number field in clear text,
+                             expressed in big-endian order.
+          16..last Data      Encrypted data for Wrap tokens with
+                             confidentiality, or plaintext data followed
+                             by the checksum for Wrap tokens without
+                             confidentiality, as described in section
+                             4.2.4.
+
+Quick notes:
+	- "EC" or "Extra Count" refers to the length of the cheksum.
+	- "Flags" (complete details in section 4.2.2) is a set of bits:
+		- if bit 0 is set, it means the token was sent by the acceptor (generally the kerberized service).
+		- bit 1 indicates that the token's payload is encrypted
+ 		- bit 2 indicates if the message is protected using a subkey defined by the acceptor.
+	- When computing checksums, EC and RRC MUST be set to 0.
+    - Wrap Tokens are not ASN.1 encoded.
+*/
+var (
+	HdrLen              = 16 // Length of the Wrap Token's header
+	GSSWrapTokenID      = [2]byte{0x05, 0x04}
+	FillerByte     byte = 0xFF
+	ChecksumECRRC       = [4]byte{0x00, 0x00, 0x00, 0x00}
+	ENC                 = binary.BigEndian
+)
+
+type WrapToken struct {
+	// const GSS Token ID: 0x0504
+	Flags byte // acceptor, sealed, acceptor subkey
+	// const Filler: 0xFF
+	EC       uint16 // checksum length. big-endian
+	RRC      uint16 // right rotation count. big-endian
+	SND_SEQ  uint64 // sender's sequence number. big-endian
+	Payload  []byte // your data! :)
+	CheckSum []byte // authenticated checksum of { payload | header }
+}
+
+// Get them bytes!
+func (wt *WrapToken) Marshal() ([]byte, error) {
+	if wt.CheckSum == nil {
+		return nil, errors.New("Checksum has not been set.")
+	}
+	if wt.Payload == nil {
+		return nil, errors.New("Payload has not been set.")
+	}
+
+	pldOffset := HdrLen                    // Offset of the payload in the token
+	chkSOffset := HdrLen + len(wt.Payload) // Offset of the checksum in the token
+
+	bytes := make([]byte, chkSOffset+int(wt.EC))
+	copy(bytes[0:], GSSWrapTokenID[:])
+	bytes[2] = wt.Flags
+	bytes[3] = FillerByte
+	ENC.PutUint16(bytes[4:6], wt.EC)
+	ENC.PutUint16(bytes[6:8], wt.RRC)
+	ENC.PutUint64(bytes[8:16], wt.SND_SEQ)
+	copy(bytes[pldOffset:], wt.Payload)
+	copy(bytes[chkSOffset:], wt.CheckSum)
+	return bytes, nil
+}
+
+func (wt *WrapToken) ComputeAndSetCheckSum(key types.EncryptionKey, keyUsage uint32) error {
+	if wt.Payload == nil {
+		return errors.New("Payload has not been set.")
```

Comment:
best practice for Go is for error messages to start with a lower case character.